File: 0547-number-of-provinces/0547-number-of-provinces.py
class UnionFind:
    def __init__(self, n):
        self.root = [i for i in range(n)]
        self.rank = [1] * n
        self.count = n

    def find(self, x):
        if x == self.root[x]:
            return x
        self.root[x] = self.find(self.root[x])
        return self.root[x]
    
    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.root[rootY] = rootX
            elif self.rank[rootY] > self.rank[rootX]:
                self.root[rootX] = rootY
            else:
                self.root[rootY] = rootX
                self.rank[rootX] += 1
            self.count -= 1
    def getCount(self):
        return self.count

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        if not isConnected or len(isConnected) == 0:
            return 0
        n = len(isConnected)
        uf = UnionFind(n)

        for row in range(n):
            for col in range(n):
                if isConnected[row][col] == 1:
                    uf.union(row, col)
        
        return uf.getCount()

File: 0547-number-of-provinces/0547-number-of-provinces.js
/**
 * @param {number[][]} isConnected
 * @return {number}
 */
var findCircleNum = function (isConnected) {
    if (isConnected.length == 0) return 0;
    let n = isConnected.length;
    const uf = new UnionFind(n);

    for (let row = 0; row < n; row++) {
        for (let col = 0; col < n; col++) {
            if (isConnected[row][col]) {
                uf.union(row, col);
            }
        }
    }

    return uf.getCount();

};

class UnionFind {
    constructor(n) {
        this.root = [...Array(n)].map((_, i) => i);
        this.rank = [...Array(n)].map((_) => 1);
        this.count = n;
    }

    find(x) {
        if (x == this.root[x]) {
            return x;
        }
        this.root[x] = this.find(this.root[x]);
        return this.root[x];
    }
    union(x, y) {
        let rootX = this.find(x);
        let rootY = this.find(y);

        if (rootX !== rootY) {
            if (this.rank[rootX] > this.rank[rootY]) {
                this.root[rootY] = rootX;
            } else if (this.rank[rootY] > this.rank[rootX]) {
                this.root[rootX] = rootY;
            } else {
                this.root[rootY] = rootX;
                this.rank[rootX] += 1;
            }
                this.count--;
        }
    }

    getCount(){
        return this.count;
    }
}


